lib/index.js

```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
 
   validateOptions(require('./options.json'), options, 'PostCSS Loader')
 
-  const rc = {
-    path: path.dirname(file),
-    ctx: {
-      file: {
-        extname: path.extname(file),
-        dirname: path.dirname(file),
-        basename: path.basename(file)
-      },
-      options: {}
-    }
-  }
-
-  if (options.config) {
-    if (options.config.path) {
-      rc.path = path.resolve(options.config.path)
+  Promise.resolve().then(() => {
+    if (options.exec || options.parser || options.syntax || options.stringifier || options.plugins) {
+      return parseOptions.call(this, options)
     }
 
-    if (options.config.ctx) {
-      rc.ctx.options = options.config.ctx
+    const rc = {
+      path: path.dirname(file),
+      ctx: {
+        file: {
+          extname: path.extname(file),
+          dirname: path.dirname(file),
+          basename: path.basename(file)
+        },
+        options: {}
+      }
     }
-  }
 
-  const sourceMap = options.sourceMap
+    // Read options from options.config only when options.config is object
+    if (options.config) {
+      if (options.config.path) {
+        rc.path = path.resolve(options.config.path)
+      }
 
-  Promise.resolve().then(() => {
-    const length = Object.keys(options).length
-
-    // TODO
-    // Refactor
-    if (!options.config && !sourceMap && length) {
-      return parseOptions.call(this, options)
-    } else if (options.config && !sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (!options.config && sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (options.config && sourceMap && length > 2) {
-      return parseOptions.call(this, options)
+      if (options.config.ctx) {
+        rc.ctx.options = options.config.ctx
+      }
     }
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
-  }).then((config) => {
-    if (!config) config = {}
+      .then((config) => {
+        // Get `sourceMap` option from loader options when no `map` option in `postcss.config.js`
+        if (!config.options.map) config.options.map = options.sourceMap
 
+        return config
+      })
+  }).then((config) => {
     if (config.file) this.addDependency(config.file)
 
+    let sourceMap = config.options.map
     let plugins = config.plugins || []
-    let options = Object.assign({
+
+    // Disable override `to` option
+    if (config.options.to) delete config.options.to
+    // Disable override `from` option
+    if (config.options.from) delete config.options.from
+    // Disable override `map` option
+    if (config.options.map) delete config.options.map
+
+    let postcssOption = Object.assign({
```

Comment:
@michael-ciniawsky because `options` related to `loader` (`const options = loaderUtils.getOptions(this) || {}`), `postcssOptions` related to `postcss` options